### PR TITLE
Multi-root workspace support

### PR DESCRIPTION
**Pull request by ritwickdey**
## PR Type

What kind of change does this PR introduce?

- Multi-root workspace support. The extension will ask the user which workspace need to be picked up. 
- No more support for single file open (without workspace).
- New Settings `multiRootWorkspaceName`
- Small refactor

```html
[ ] Bugfix
[x] Feature
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other: <!-- Please describe: -->
```

## What is the current behavior?

Issue Number: #43 

## What is the new behavior?

```text
[ ] Yes
[x] No (Though Not sure :p )
```

### CODE REVIEW

File: src/appModel.ts [hunk 1]
```diff
@@ -141,20 +152,13 @@ export class AppModel {
...
-            if (!textEditor) return;
-
-            // If a HTML file open without Workspace
-            if (workspace.rootPath === undefined && Helper.IsSupportedFile(textEditor.document.fileName)) {
-                return callback();
-            }
+    private haveAnySupportedFile() {
+        return new Promise<void>(resolve => {
+            const globFormat = `**/*[${SUPPRORTED_EXT.join(' | ')}]`;
+            workspace.findFiles(globFormat, '**/node_modules/**', 1)
```
Comment (rjoydip) [inline src/appModel.ts]: Configure ignore like this and pass `config` which is coming from `settings(user settings)` in below `ignore` array.

```js
const ignore = [
    'build/**/*',
    'out/**/*',
    'dist/**/*',
];
const workspaceFiles = await workspace.findFiles(
    new RelativePattern(workspaceFolder, `{${ignoreWorkspace.join(',')}}`),
);
```

Comment (rjoydip) [reply src/appModel.ts]: Also, there should be `catch` if something went wrong.

Comment (ritwickdey) [reply src/appModel.ts]: Sorry, I don't get that. 

BTW, `workspace.findFiles` returns `Thenable<T>`, we can't use `await`.
And `haveAnySupportedFile` was introduced at early stage of `live server` just for not to visible  the status bar 'go live' button if there're no SUPPORTED file (html, svg & htm) in the workspace folder - though it's not working well.

File: src/workspaceResolver.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { workspace, window } from 'vscode';
+import { Config } from './Config';
+
+
+export const workspaceResolver = () => {
```
Comment (rjoydip) [inline src/workspaceResolver.ts]: export function workspaceResolver() { /* ... */ }

File: src/workspaceResolver.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { workspace, window } from 'vscode';
+import { Config } from './Config';
+
+
+export const setOrChangeWorkspace = () => {
```
Comment (rjoydip) [inline src/workspaceResolver.ts]: export function workspaceResolver () { /* ... */ }

File: src/workspaceResolver.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        ignoreFocusOut: true
+    }).then(workspaceName => {
+        if (workspaceName) {
+            return Config.setMutiRootWorkspaceName(workspaceName).then(() => workspaceName);
+        }
+    });
+};
+
+
+export const workspaceResolver = () => {
```
Comment (rjoydip) [inline src/workspaceResolver.ts]: export function workspaceResolver() { /* ... */ }

File: src/workspaceResolver.ts (hunk 3 above)
Comment (ritwickdey) [reply src/workspaceResolver.ts]: Okay! I've no issue. Is there any reason that `function literals` is bad?

Comment (rjoydip) [reply src/workspaceResolver.ts]: Not an issue but for better readblity

File: src/appModel.ts (hunk 1 above)
Comment (rjoydip) [reply src/appModel.ts]: Please check this https://github.com/Microsoft/vscode/issues/37806

File: src/workspaceResolver.ts (hunk 3 above)
Comment (ritwickdey) [reply src/workspaceResolver.ts]: Okk. okk

File: src/appModel.ts (hunk 1 above)
Comment (ritwickdey) [reply src/appModel.ts]: yep! I was also thinking that it is a bug. Anyway, this is not the main focus - it is just about user experience that the 'Go Live' will be visible if and only there has any HTML, HTM or SVG file.

**Comment by ritwickdey** [comment]
Still, edge cases are not handled.

File: src/appModel.ts (hunk 1 above)
Comment (rjoydip) [reply src/appModel.ts]: Yes that's good. Go live should not visible for all types of files. It's really confusing

**Comment by ritwickdey** [comment]
@rjoydip, yep test is now really important. But testing vscode extension is hard i think. I'll try to add tests for Helper Class. I don't know how I can test the GoLive Method or Live Server Helper class or change detection of HTML.

**Comment by rjoydip** [comment]
I have to try I haven't written any test cases in vscode.

**Comment by rjoydip** [comment]
Yes you can ship this. LGTM

**Comment by champolot** [comment]
Hi there, I installed live server but there's no live server button below the coding area and there it says that theres no live server installed. Is there is any way to debug this? 

image

**Pull request merged**